account_invoice_analytic_line/models/account_analytic_line.py

```diff
@@ -0,0 +1,112 @@
+                lines['__domain']
+            )._account_invoice_analytic_line_create_invoice()
+        action = self.env['ir.actions.act_window'].for_xml_id(
+            'account', 'action_invoice_tree1',
+        )
+        action['domain'] = [('id', 'in', invoices.ids)]
+        return action
+
+    @api.multi
+    def _account_invoice_analytic_line_create_invoice(self):
+        """Actually create an invoice for lines. We assume they have the same
+        partner"""
+        if not self:
+            return self.env['account.invoice']
+        first = self[:1]
+        invoice_vals = {
+            'partner_id':
+            first.partner_id.commercial_partner_id.address_get(['invoice'])
+            .get('invoice', first.partner_id.commercial_partner_id.id),
+            'type': 'out_invoice',
+        }
+        invoice_vals = self.env['account.invoice'].play_onchanges(
+            invoice_vals, ['partner_id'],
+        )
+        invoice = self.env['account.invoice'].create(invoice_vals)
+        for lines in self._account_invoice_analytic_line_group_invoice_lines():
+            self.search(
+                lines['__domain']
+            )._account_invoice_analytic_line_create_invoice_line(invoice)
+        invoice.compute_taxes()
+        return invoice
+
+    @api.multi
+    def _account_invoice_analytic_line_create_invoice_line(self, invoice):
+        """Actually create an invoice line from lines in self. We assume they
+        have the same product, project and discount"""
+        if not self:
+            return self.env['account.invoice.line']
+        first = self[:1]
+        quantity = sum(self.mapped('unit_amount'))
+        product = (
+            first.product_id or
+            first.account_id.account_invoice_analytic_line_product_id or
+            invoice.partner_id.account_invoice_analytic_line_product_id
+        )
+        if not product:
+            raise UserError(_(
+                'Neither your line, nor the project nor the partner define a '
+                'product to use for invoicing, you need to set one of them.'
+            ))
+        invoice_line_vals = {
+            'product_id': product.id,
+            'invoice_id': invoice.id,
+            'discount': first.invoice_discount_id.discount,
+            'quantity': quantity,
+            'price_unit': product.with_context(
+                pricelist=invoice.partner_id.property_product_pricelist.id,
+                partner=invoice.partner_id.id,
+                quantity=quantity,
+            ).price,
+        }
+        invoice_line_vals = self.env['account.invoice.line'].play_onchanges(
+            invoice_line_vals,
+            ['product_id', 'price_unit', 'discount', 'quantity'],
+        )
+        invoice_line_vals['name'] = first.project_id.name
+        invoice_line = self.env['account.invoice.line'].create(
+            invoice_line_vals,
+        )
+        self.write({'invoice_line_id': invoice_line.id})
```

Comment:
Why do we need this write?

Reply:
note that `self` contains the analytic lines we've just created the invoice line for. So this takes care that we have a link between the analytic lines and the invoice line created. This is used to determine which analytic lines are not yet invoiced, for example in https://github.com/OCA/account-invoicing/pull/462/files/ad8bc705b3d7e279a2bdbf75d26dc2641d856d71#diff-c1268f2774a7fdc6bcc96f5e1bcedee9R21